### PR TITLE
Various fixes to Fierce boefje

### LINKED ISSUE #904: Fierce boefje: subdomains get L0 indemnification
**Describe the bug**
When scanning `example.com` with KAT at L4 the domain `subdomain.example.com` (sometimes) gets L0 as indemnification level, while sometimes it gets L1. There may be a difference in using `Fierce` or `DnsRecords`/`Subfinder` here.

**To Reproduce**
1. Scan (fierce?) a domain with a subdomain
2.  Notice the indemnification of the subdomain

**Expected behavior**
I would expect at least L1 for a subdomain of the first entered domain in KAT.

**OpenKAT version**
1.8rc2

### PR DESCRIPTION
**Pull request by praseodym**
### Changes

**Fix error handling in Fierce boefje** 
Connection errors in HTTP HEAD requests were not handled properly.

**Change Fierce normalizer to ResolvedHostname** 
Fierce returns the resolved hostname, not a DNS record. This caused
incorrect OOIs for CNAME records.

**Add registered_domain to Hostname model and Fierce boefje**

### Issue link
Fixes #904

### Proof
n/a

---

### Code Checklist
- [x] All the commits in this PR are properly PGP-signed and verified;
- [x] This PR only contains functionality relevant to the issue; tickets have been created for newly discovered issues.
- [x] I have written unit tests for the changes or fixes I made.
- [ ] For any non-trivial functionality, I have added integration and/or end-to-end tests.
- [ ] I have performed a self-review of my code and refactored it to the best of my abilities.

### Communication
- [ ] I have informed others of any required `.env` changes files if required and changed the `.env-dist` accordingly.
- [ ] I have made corresponding changes to the documentation, if necessary.

---
## Checklist for code reviewers:
Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---
## Checklist for QA:
Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.

### CODE REVIEW

**Comment by Darwinkel** [comment]
## Checklist for QA:
- [x] I have checked out this branch, and successfully ran a fresh `make reset`.
- [x] I confirmed that there are no unintended functional regressions in this branch:
    - [x] I have managed to pass the onboarding flow
    - [x] Objects and Findings are created properly
    - [x] Tasks are created and completed properly
- [x] I confirmed that the PR's advertised `feature` or `hotfix` works as intended.

### What works:
* Subdomain clearance levels are properly inherited to L2 when using Fierce
* No apparent regressions

**Comment by underdarknl** [comment]
Seeing that we are now doing http requests against the hosts we find, should we maybe also store the http headers and port so we can create ipservices based on the port, and also create the subsequent http headers? It should be a small step to at least include them in the output of the boefje. (possibly, we can store them in the same mime-type the http-request boefje stores its data, so that normalizer will pick up the data).

**Comment by praseodym** [comment]
> Seeing that we are now doing http requests against the hosts we find, should we maybe also store the http headers and port so we can create ipservices based on the port, and also create the subsequent http headers? It should be a small step to at least include them in the output of the boefje. (possibly, we can store them in the same mime-type the http-request boefje stores its data, so that normalizer will pick up the data).

The HTTP requests aren't new, Fierce was always doing those. This PR just fixes the error handling for cases where a connection cannot be made. 

Fierce doesn't return any output from those HTTP requests, so unless we make more changes to it we can't parse those into OOIs. We can pick that up in a new issue if desired.

File: octopoes/tests/conftest.py [hunk 1]
```diff
@@ -72,7 +72,8 @@ def list_neighbours(self, references: Set[Reference], paths: Set[Path], valid_ti
             if segment.direction == Direction.OUTGOING:
                 for ref in references:
                     neighbour_ref = getattr(self.oois[ref], segment.property_name)
-                    neighbours.add(self.oois[neighbour_ref])
+                    if neighbour_ref is not None:
+                        neighbours.add(self.oois[neighbour_ref])
```
Comment (ammar92) [inline octopoes/tests/conftest.py:76]: Nice find

**Review by ammar92 — APPROVED** [review]
Looks good, no remarks

File: boefjes/boefjes/plugins/kat_fierce/normalize.py [hunk 2]
```diff
@@ -16,23 +17,22 @@ def run(normalizer_meta: NormalizerMeta, raw: Union[bytes, str]) -> Iterable[OOI
...
-            )
+        hostname = subdomain["url"].rstrip(".")
+        registered_domain = tldextract.extract(hostname).registered_domain
+
+        registered_domain_ooi = Hostname(name=registered_domain, network=internet.reference)
+        yield registered_domain_ooi
+        hostname_ooi = Hostname(
+            name=hostname, network=internet.reference, registered_domain=registered_domain_ooi.reference
+        )
+        yield hostname_ooi
```
[the remark this answers is not in the document]
Comment (praseodym) [reply boefjes/boefjes/plugins/kat_fierce/normalize.py:28]: I'd already thought of adding `registered_domain` in more places after this PR. I think there are two options:
1. Amend all normalizers to add `registered_domain`
2. Do this enrichment in a central place (do we currently have such a place?)

What do you think?

[the remark this answers is not in the document]
Comment (Lisser) [reply boefjes/boefjes/plugins/kat_fierce/normalize.py:28]: Regarding option 2: 
Yes, a Bit would be the appropriate place you are looking for. Bits are meant to derive logical conclusions from premises. (In this instance, a Hostname representing a subdomain of a registered domain, logically implies the existance of that domain and their relationship.)

However, I believe that current restrictions in Octopoes do not allow Bits to mutate its source object. As long as this is not resolved, option 1 is the best way to go forward with this. 

Could you verify this @noamblitz ?

[the remark this answers is not in the document]
Comment (noamblitz) [reply boefjes/boefjes/plugins/kat_fierce/normalize.py:28]: Bits are able to change the attributes of the input ooi that are not part of the primary key.

**Review by underdarknl — APPROVED** [review]

File: boefjes/boefjes/plugins/kat_fierce/normalize.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (praseodym) [reply boefjes/boefjes/plugins/kat_fierce/normalize.py:28]: Created #1037 to add a Bit for this.

**Pull request merged**